pandas/core/arrays/boolean.py

```diff
@@ -740,6 +742,172 @@ def boolean_arithmetic_method(self, other):
+    #     return result, mask
+
+    result = left | right
+    mask[left & ~left_mask] = False
+    if right_mask is not None:
+        mask[right & ~right_mask] = False
+    elif right is True:
+        mask[:] = False
+
+    # update
+    return result, mask
+
+
+def kleene_xor(
+    left: Union[bool, np.ndarray],
+    right: Union[bool, np.ndarray],
+    left_mask: Optional[np.ndarray],
+    right_mask: Optional[np.ndarray],
+):
+    """
+    Boolean ``xor`` using Kleene logic.
+
+    This is the same as ``or``, with the following adjustments
+
+    * True, True -> False
+    * True, NA   -> NA
+
+    Parameters
+    ----------
+    left, right : ndarray, NA, or bool
+        The values of the array.
+    left_mask, right_mask : ndarray, optional
+        The masks. When
+
+    Returns
+    -------
+    result, mask: ndarray[bool]
+        The result of the logical xor, and the new mask.
+    """
+    if left_mask is None:
+        return kleene_xor(right, left, right_mask, left_mask)
+
+    raise_for_nan(right, method="xor")
+    # Re-use or, and update with adjustments.
+    result, mask = kleene_or(left, right, left_mask, right_mask)
+
+    # # TODO(pd.NA): change to pd.NA
+    # if lib.is_scalar(right) and right is libmissing.NA:
+    #     # True | NA == True
+    #     # True ^ NA == NA
+    #     mask[result] = True
+
+    result[left & right] = False
+    mask[right & left_mask] = True
+    if right_mask is not None:
+        mask[left & right_mask] = True
+
+    result[mask] = False
+    return result, mask
+
+
+def kleene_and(
+    left: Union[bool, np.ndarray],
+    right: Union[bool, np.ndarray],
+    left_mask: Optional[np.ndarray],
+    right_mask: Optional[np.ndarray],
+):
+    """
+    Boolean ``and`` using Kleene logic.
+
+    Values are ``NA`` for ``NA & NA`` or ``True & NA``.
+
+    Parameters
+    ----------
+    left, right : ndarray, NA, or bool
+        The values of the array.
+    left_mask, right_mask : ndarray, optional
+        The masks. When
+
+    Returns
+    -------
+    result, mask: ndarray[bool]
+        The result of the logical xor, and the new mask.
+    """
+    if left_mask is None:
+        return kleene_and(right, left, right_mask, left_mask)
+
+    assert isinstance(left, np.ndarray)
+    raise_for_nan(right, method="and")
+    mask = left_mask
+
+    if right_mask is not None:
+        mask = mask | right_mask
+    else:
+        mask = mask.copy()
+
+    if lib.is_scalar(right):
+        result = left.copy()
+        mask = left_mask.copy()
+        if np.isnan(right):
+            # TODO(pd.NA): change to NA
+            mask[result] = True
+        else:
+            result = result & right  # already copied.
+            if right is False:
+                # unmask everything
+                mask[:] = False
+    else:
+        result = left & right
+        # unmask where either left or right is False
+        mask[~left & ~left_mask] = False
+        mask[~right & ~right_mask] = False
```

Comment:
Similar comment here, I think that something like this can be faster:

```
left_false = ~left & ~left_mask
right_false= ~right & ~right_mask

mask = (left_mask & ~right_false) | (right_mask & ~left_false)
```

(avoiding setitem)

And need to think if we can avoid some `~`

Reply:
Further optimization:

```
left_false = ~(left | left_mask)
right_false= ~(right | right_mask)
mask = (left_mask & ~right_false) | (right_mask & ~left_false)
```

Timing comparison:

```
left = np.random.randint(0, 2, 1000).astype(bool)
right = np.random.randint(0, 2, 1000).astype(bool) 
left_mask = np.random.randint(0, 2, 1000).astype(bool) 
right_mask = np.random.randint(0, 2, 1000).astype(bool)
```

```
In [47]: %%timeit 
    ...: mask = left_mask | right_mask 
    ...: mask[~left & ~left_mask] = False 
    ...: mask[~right & ~right_mask] = False 

7.2 µs ± 106 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [58]: %%timeit 
    ...: left_false = ~(left | left_mask) 
    ...: right_false= ~(right | right_mask) 
    ...:  
    ...: mask = (left_mask & ~right_false) | (right_mask & ~left_false)
3.73 µs ± 275 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```

Reply:
And on bigger arrays, the difference is much bigger, it seems. For 100_000 elements, I get 775 µs vs 45 µs

Reply:
Thanks. I'll also add an asv for these ops.